### PR TITLE
 Implemented Game Asset Registry Smart Contract

### LINKED ISSUE #52: [Contract] 📂 Implement Game Asset Registry
## 🎯 Objective

Create a complete smart contract for `Game Asset Registry` that allows users to `register, own, and track in-game assets` with full `metadata support and history tracking`.

---

## 🏗 Contract structure

```rust
game-asset-registry/src/
  lib.rs          // Configuration and contract exports
  registry.rs     // Core logic for asset registration and metadata handling
  ownership.rs    // Handles ownership transfer and validation
  history.rs      // Stores and retrieves item history data
```

---

## 🗂 Requirements

1. **Game Asset Functionality**
   - Register unique in-game items (NFT-like)
   - Assign and transfer ownership
   - Store and update item metadata (e.g., stats, type, rarity)
   - Track ownership and item history
   - Query asset details on-chain

---

## 📦 Key Data Structures

```rust
struct GameAsset {
    id: BytesN<32>,
    owner: Address,
    metadata: AssetMetadata,
    history: Vec<OwnershipRecord>,
}

struct AssetMetadata {
    name: String,
    item_type: String,
    attributes: Vec<(String, String)>, // e.g., ("attack", "15"), ("rarity", "epic")
}

struct OwnershipRecord {
    previous_owner: Address,
    timestamp: u64,
    reason: String, // e.g., "initial mint", "traded", "rewarded"
}
```

---

## 🔑 Key Functions

- `register_asset()` – Mint a new unique asset with metadata
- `transfer_asset()` – Transfer ownership to another player
- `update_metadata()` – Update asset attributes and stats
- `get_asset_info()` – Retrieve full details and metadata of an asset
- `get_asset_history()` – Return the ownership history of an asset

---

## 🔗 References

- **[Stellar Official Guide](https://developers.stellar.org/#smart-contract-developers)**
- **[Rust Book](https://doc.rust-lang.org/book/)**

---

## ⏳ ETA 
- 2 Days

---

## 📌 Additional Notes

- Consider compatibility with game engines (e.g., Unity/WebGL)
- Prioritize low gas cost design for frequent asset updates
- Ensure secure and verifiable ownership records
- Potential future integration with marketplaces or tournaments

### PR DESCRIPTION
**Pull request by manassehO**
# PR: Implement Game Asset Registry Smart Contract

## Overview
This PR implements a complete Game Asset Registry smart contract that enables the registration, ownership, and tracking of in-game assets with full metadata support and ownership history.

## Key Features
- **Asset Registration**: Create unique in-game items with arbitrary metadata
- **Ownership Management**: Transfer assets between players with full history tracking
- **Metadata Management**: Update and modify asset attributes and properties
- **Comprehensive Queries**: APIs for retrieving assets by ID, owner, and pagination

## Implementation Details
- **Data Structures**:
  - `GameAsset`: Core asset structure with ID, owner, metadata, and history
  - `AssetMetadata`: Flexible metadata structure with key-value attribute pairs
  - `OwnershipRecord`: History record for ownership transfers

- **Core Functions**:
  - `register_asset()`: Create new assets with initial owner and metadata
  - `transfer_asset()`: Transfer ownership between players with reason tracking
  - `update_metadata()`: Update asset attributes (by owner or admin)
  - `get_asset_info()`: Retrieve full asset details
  - `get_asset_history()`: Get ownership history of an asset
  - `get_user_assets()`: List all assets owned by a user
  - `get_asset_count()`: Get total number of registered assets

- **Security**:
  - Admin-only asset registration **(Do we want only admin to register assets?)**
  - Owner-only asset transfers
  - Owner or admin metadata updates
  - Immutable ownership history

## Testing
Implemented comprehensive test suite covering:
- Asset registration and validation
- Ownership transfers and authorization checks
- Metadata updates and validation
- Query functions and pagination
- Error handling and edge cases

Screenshot 2025-05-01 at 12 41 23 PM

- closes #52

### CODE REVIEW

File: apps/contracts/contracts/asset-registry/cargo.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "asset-registry"
+version = "0.0.1"
```
Comment (aguilar1x) [inline apps/contracts/contracts/asset-registry/cargo.toml:3]: change the versión to 0.1.0

**Review by aguilar1x — CHANGES_REQUESTED** [review]
Delete the test file and snapshots of the test, since there's already an issue created for their implementation.

**Comment by manassehO** [comment]
done

**Review by aguilar1x — CHANGES_REQUESTED** [review]
Fix the conflicts in your branch

**Review by aguilar1x — CHANGES_REQUESTED** [review]
Do not delete the game-review contract files.

**Comment by manassehO** [comment]
oops,. that wasn't intentional.

reverting

**Comment by manassehO** [comment]
Fixed @aguilar1x
